### ISSUE TITLE
Bytecode interpreter extends boolean to uint32 in JNI calls

### ISSUE THREAD

**Issue by knn-k** [eclipse-openj9/openj9#14574]
PR #4569 introduced a change in the bytecode interpreter that extends `boolean` arguments to `uint32` in JNI calls, as a workaround for a type inconsistency bug in Issue #4193.

You must not do that on AArch64 macOS.  The native type for `jboolean` is `unsigned char`, and the JNI code allocates only 1-byte slot in the C stack for a `jboolean` argument on that platform.

I open this Issue to discuss whether the workaround is still needed or not.

**Comment by knn-k** [comment]
AArch64 Linux works with the type extension because its C stack usage is different from that on AArch64 macOS.
See also the discussion in #14500 and #14560.

**Comment by ChengJin01** [comment]
FYI:  @gacholio, @DanHeidinga, @tajila, @pshipton

**Comment by knn-k** [comment]
I opened PR #14575 as a temporary change for AArch64 macOS.

**Comment by knn-k** [comment]
I want to fix the problem around JNI arguments on AArch64 macOS in JDK 18 that will be released next month.

- Stop extending boolean only on AArch64 macOS: Apply the change #14575 to the master branch and to the 0.31.0 branch.
- Stop extending boolean on all the platforms: I guess it is too late for doing this in the 0.31.0 branch.
- Any other idea?

**Comment by knn-k** [comment]
> how to address this case when the conversion for `boolean` to `jint` unfortunately occurs in user applications which work good with Hotspot in the future

Well, the JNI code must be fixed correctly in that case, not adding workaround in the VM side, I believe.

What about other data types, such as `jbyte`, `jshort`, and `jchar`?  Why not extend all of them to `jint`?

**Comment by ChengJin01** [comment]
> What about other data types, such as jbyte, jshort, and jchar? Why not extend all of them to jint?

It totally depends on whether it happens in OpenJDK/user applications (So far we only spotted the conversion from `boolean` to `jint` in the past).  So we might need to deal with that if any test shows it works that way on Hotspot.

Strictly speaking, such conversion is incorrect and should be avoided but we need our solution to work through if it happens in which case there is no way to modify the existing applications. That's why we created the fix at https://github.com/eclipse-openj9/openj9/pull/4569 on VM side even though we proposed a bug request for OpenJDK, assuming they would simply ignore that given the application works fine on Hotspot.

**Comment by knn-k** [comment]
No worries.
I wrote a small JNI program, and it shows that Hotspot VM fails on AArch64 macOS when there are inconsistent type declarations (`boolean` in the Java code vs `jint` in the native code).
We will be able to say "Fix the JNI code" by showing the result with Hotspot on AArch64 macOS.

**Comment by knn-k** [comment]
Attaching the test code I mentioned above.  [JNITest.zip](https://github.com/eclipse-openj9/openj9/files/8121492/JNITest.zip)
Hotspot for AArch64 macOS shows the following result, where `g` and `h` are `boolean` in the Java code side:

```
% hotspot/jdk-17.0.2+8/Contents/Home/bin/java -Djava.library.path=. JNITest
g=1868759041 h=1234567 i=88492216
result = 88492216
```

**Comment by ChengJin01** [comment]
What about the result from Hotspot on Linux/Aarch64?

**Comment by knn-k** [comment]
Hotspot for AArch64 Linux does not extend `boolean` to 4 bytes when passing it to the JNI function as the following output shows.  The value for `i` is correct here, because the C stack usage is different between macOS and Linux:

```
$ hotspot/jdk-17.0.2+8/bin/java -Djava.library.path=. JNITest
g=2021655297 h=-1811795968 i=1234567
result = 1234567
```

**Comment by ChengJin01** [comment]
Why the boolean arguments start from 7 and 8 in the test? Any special reason for that ?  Probably better have different combination of arguments to verify the results.

@gacholio, @DanHeidinga, any input/concern as to whether we still need the PR at https://github.com/eclipse-openj9/openj9/pull/4569 (Just reminder that the PR works good on all platforms except macOS/Aarch64).

**Comment by gacholio** [comment]
Can someone explain the actual problem here?

**Comment by knn-k** [comment]
The reason is simple.  I wanted to show how the `boolean` arguments are passed in the C stack.  AArch64 uses 8 GPRs for passing arguments.  The JNI code uses two GPRs `x0` and `x1` for `JNIEnv*` and `jobject`, and 6 arguments `a`-`f` are passed in registers `x2`-`x7`.
The remaining arguments `g`-`i` are passed via the C stack, and the layout is as follows when the type declarations are correct:

AArch64 Linux:
```
g: at offset n
h: at offset n+8
i: at offset n+16
```

AArch64 macOS (no space for storing 4-byte values for `g` and `h`):
```
g: at offset n
h: at offset n+1
i: at offset n+4
```

**Summary**
We must not extend `boolean` to 4 bytes long.
Let's stop doing it in OpenJ9 bytecode interpreter.

**Comment by knn-k** [comment]
@gacholio

**Facts**:
- OpenJ9 for AArch64 macOS won't work as expected with correct JNI functions if the VM extends `boolean` arguments to 4-byte wide.  I have proved it and explained the reason.
- Hotspot does not do the extension either on AArch64 macOS or AArch64 Linux.  I have shown it in my comments above.

**My message**:
- I need to revert the change from #4569 at least for AArch64 macOS in v.0.31 next month.  I have opened PR #14575 for that.
- I also want to stop extending `boolean` in JNI calls on all other platforms in the near future if the workaround is only for a specific problem in the past.

**My question**:
- What would be the problem if we throw away the workaround on all the platforms?  @ChengJin01 explained in his comments above, but I am not convinced.
